polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -430,353 +452,207 @@ impl Scope {
-				.nodes
-				.iter()
-				.take_while(|n| n.parent == NodePointer::Root)
-				.enumerate()
-				.find(|(_, n)| &n.candidate_hash == candidate_hash)
-				.map(|(i, _)| NodePointer::Storage(i)),
-			NodePointer::Storage(ptr) =>
-				self.nodes.get(ptr).and_then(|n| n.candidate_child(candidate_hash)),
-		}
-	}
-
 	/// Returns an O(n) iterator over the hashes of candidates contained in the
 	/// tree.
 	pub(crate) fn candidates(&self) -> impl Iterator<Item = CandidateHash> + '_ {
-		self.candidates.keys().cloned()
+		self.candidates.iter().cloned()
 	}
 
-	/// Whether the candidate exists and at what depths.
-	pub(crate) fn candidate(&self, candidate: &CandidateHash) -> Option<Vec<usize>> {
-		self.candidates.get(candidate).map(|d| d.iter_ones().collect())
+	/// Returns the number of candidates in the chain
+	pub(crate) fn len(&self) -> usize {
+		self.candidates.len()
 	}
 
-	/// Add a candidate and recursively populate from storage.
-	///
-	/// Candidates can be added either as children of the root or children of other candidates.
-	pub(crate) fn add_and_populate(&mut self, hash: CandidateHash, storage: &CandidateStorage) {
-		let candidate_entry = match storage.get(&hash) {
-			None => return,
-			Some(e) => e,
-		};
-
-		let candidate_parent = &candidate_entry.candidate.persisted_validation_data.parent_head;
-
-		// Select an initial set of bases, whose required relay-parent matches that of the
-		// candidate.
-		let root_base = if &self.scope.base_constraints.required_parent == candidate_parent {
-			Some(NodePointer::Root)
-		} else {
-			None
-		};
-
-		let non_root_bases = self
-			.nodes
-			.iter()
-			.enumerate()
-			.filter(|(_, n)| {
-				n.cumulative_modifications.required_parent.as_ref() == Some(candidate_parent)
-			})
-			.map(|(i, _)| NodePointer::Storage(i));
-
-		let bases = root_base.into_iter().chain(non_root_bases).collect();
-
-		// Pass this into the population function, which will sanity-check stuff like depth,
-		// fragments, etc. and then recursively populate.
-		self.populate_from_bases(storage, bases);
+	/// Whether the candidate exists.
+	pub(crate) fn contains_candidate(&self, candidate: &CandidateHash) -> bool {
+		self.candidates.contains(candidate)
 	}
 
-	/// Returns `true` if the path from the root to the node's parent (inclusive)
-	/// only contains backed candidates, `false` otherwise.
-	fn path_contains_backed_only_candidates(
-		&self,
-		mut parent_pointer: NodePointer,
-		candidate_storage: &CandidateStorage,
-	) -> bool {
-		while let NodePointer::Storage(ptr) = parent_pointer {
-			let node = &self.nodes[ptr];
-			let candidate_hash = &node.candidate_hash;
-
-			if candidate_storage.get(candidate_hash).map_or(true, |candidate_entry| {
-				!matches!(candidate_entry.state, CandidateState::Backed)
-			}) {
-				return false
-			}
-			parent_pointer = node.parent;
-		}
+	/// Return a vector of the chain's candidate hashes, in-order.
+	pub(crate) fn chain(&self) -> Vec<CandidateHash> {
```

Comment:
Isn't `as_vec()` better since this is a method of `FragmentChain` and implies it is a chain already.

Reply:
renamed to `to_vec()`